src/ivc/fold_relaxed_plonk_instance_chip.rs

```diff
@@ -0,0 +1,693 @@
+
+    use poseidon::Spec;
+
+    use crate::{plonk::TableData, poseidon::poseidon_circuit::PoseidonChip};
+
+    use super::*;
+
+    #[test]
+    fn generate_challenge() {
+        let mut rnd = rand::thread_rng();
+
+        let chip = FoldRelaxedPlonkInstanceChip::<C1>::from_instance(
+            PlonkInstance {
+                W_commitment: C1::random(&mut rnd),
+                instance: vec![ScalarExt::random(&mut rnd), ScalarExt::random(&mut rnd)],
+                challenges: vec![<C1 as CurveAffine>::ScalarExt::random(&mut rnd)],
+            },
+            NonZeroUsize::new(64).unwrap(),
+            NonZeroUsize::new(4).unwrap(),
+        )
+        .unwrap();
+
+        let mut td = TableData::new(10, vec![]);
+        let mut shape: Box<dyn RegionLayouter<_>> =
+            Box::new(RegionShape::new(RegionIndex::from(0)));
+        let shape_mut: &mut dyn RegionLayouter<_> = shape.as_mut();
+
+        let region = Region::from(shape_mut);
+
+        let config = MainGate::<Base, 5>::configure(&mut td.cs);
+        let spec = Spec::<Base, 5, 4>::new(10, 10);
+
+        let _assigned_challenge = chip
+            .assign_witness_with_challenge(
+                &mut RegionCtx::new(region, 0),
+                &config.clone(),
+                PoseidonChip::new(config, spec),
+                &Base::random(&mut rnd),
+                &PlonkInstance {
+                    W_commitment: C1::random(&mut rnd),
+                    instance: vec![ScalarExt::random(&mut rnd), ScalarExt::random(&mut rnd)],
+                    challenges: vec![<C1 as CurveAffine>::ScalarExt::random(&mut rnd)],
+                },
+                &vec![
+                    C1::random(&mut rnd),
+                    C1::random(&mut rnd),
+                    C1::random(&mut rnd),
+                ],
+            )
+            .unwrap();
+    }
+
+    #[test_log::test]
+    fn fold() {
+        debug!("start");
+        const T: usize = 6;
+        let mut rnd = rand::thread_rng();
+
+        let chip = FoldRelaxedPlonkInstanceChip::<C1>::from_instance(
+            PlonkInstance {
+                W_commitment: C1::random(&mut rnd),
+                instance: vec![ScalarExt::random(&mut rnd), ScalarExt::random(&mut rnd)],
+                challenges: vec![<C1 as CurveAffine>::ScalarExt::random(&mut rnd)],
+            },
+            NonZeroUsize::new(64).unwrap(),
+            NonZeroUsize::new(256).unwrap(),
+        )
+        .unwrap();
+
+        let mut td = TableData::new(10, vec![]);
+        let mut shape: Box<dyn RegionLayouter<_>> =
+            Box::new(RegionShape::new(RegionIndex::from(0)));
+        let shape_mut: &mut dyn RegionLayouter<_> = shape.as_mut();
+
+        let region = Region::from(shape_mut);
+
+        let config = MainGate::<Base, T>::configure(&mut td.cs);
+        let spec = Spec::<Base, T, { T - 1 }>::new(10, 10);
+
+        let _fold_result = chip
+            .fold::<T>(
```

Comment:
can we compare fold result from the NIFS verifier or the `fold` function defined in `src/plonk/mod.rs`? so that we make sure the core component is correct before put everything together

Reply:
Yes, this is a part of my test strategy for this component.